Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs

```diff
@@ -970,9 +968,26 @@ private bool ExecuteOnAllServicesReverseOrder(Action<IMixedRealityService> execu
         private bool ExecuteOnAllServices(IEnumerable<IMixedRealityService> services, Action<IMixedRealityService> execute)
         {
             if (!HasProfileAndIsInitialized) { return false; }
+
+            // Get the count of services, so that we can protect against the count changing during iteration.
+            int serviceCount = -1;
+            // The MixedRealityServiceRegistry returns the service collection as an IReadOnlyCollection
+            IReadOnlyCollection<IMixedRealityService> serviceCollection = services as IReadOnlyCollection<IMixedRealityService>;
+            if (serviceCollection != null)
+            {
+                serviceCount = serviceCollection.Count();
+            }
+
             foreach (var system in services)
             {
                 execute(system);
+
+                // If we successfully got the count of services and the count has changed, exit the loop.
+                if ((serviceCount != -1) &&
+                    (serviceCount != MixedRealityServiceRegistry.GetAllServices().Count()))
+                {
+                    break;
+                }
```

Comment:
It seems like your aim here is just to iteration through an IEnumerable<IMixedRealityService> while the collection might be getting modified (for example, if you changed profiles and now a service no longer exists.

Did you consider something like this:
```
using (IEnumerator<IMixedRealityService> enumerator = services.GetEnumerator())
{
    bool moveNext = enumerator.MoveNext();
    while (moveNext)
    {
        execute(enumerator.Current);
        moveNext = enumerator.MoveNext();
    }
}
```
From [this source](https://blog.magnusmontin.net/2015/02/28/enumerating-collections-that-change-in-csharp/). That has advantage of still enumerating over all services that exist, and also is a bit easier to understand (I had to think a bit about why you were doing this count comparison).

Reply:
i'll give your example a try.

Reply:
This would also be nice to be able to skip the MRTKRegistry.GetAllServices<>() call